functions/go/set-image/transformer/images.go

```diff
@@ -0,0 +1,237 @@
+package transformer
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+	"sigs.k8s.io/kustomize/api/image"
+	"sigs.k8s.io/kustomize/api/types"
+)
+
+const FnConfigKind = "SetImage"
+
+type FieldPath []string
+
+type Image struct {
+	// Name is a tag-less image name. should be deprecate, means image name
+	Name string `json:"name,omitempty" yaml:"name,omitempty"`
+
+	// ContainerName is the name for container
+	ContainerName string `json:"containerName,omitempty" yaml:"containerName,omitempty"`
+
+	// ImageName is the image name
+	ImageName string `json:"imageName,omitempty" yaml:"imageName,omitempty"`
+
+	// NewName is the value used to replace the original name, replace image name
+	NewName string `json:"newName,omitempty" yaml:"newName,omitempty"`
+
+	// NewTag is the value used to replace the original tag.
+	NewTag string `json:"newTag,omitempty" yaml:"newTag,omitempty"`
+
+	// Digest is the value used to replace the original image tag.
+	// If digest is present NewTag value is ignored.
+	Digest string `json:"digest,omitempty" yaml:"digest,omitempty"`
+}
+
+// ImageTransformer supports the set-image workflow, it uses Config to parse functionConfig, Transform to change the image
+type ImageTransformer struct {
+	// Image is the desired image
+	Image Image
+	// AdditionalImageFields is used to specify additional fields to set image
+	AdditionalImageFields types.FsSlice
+	// Results logs the changes to the KRM resource image
+	Results fn.Results
+	// ResultCount logs the total count image change
+	ResultCount int
+}
+
+// Config parse the functionConfig kubeObject to the fields in the ImageTransformer
+func (imageTrans *ImageTransformer) Config(functionConfig *fn.KubeObject) error {
+	switch {
+	case functionConfig.IsEmpty():
+		return fmt.Errorf("Config is Empty, failed to configure function: `functionConfig` must be either a `ConfigMap` or `SetImage`")
+	case functionConfig.IsGVK("", "v1", "ConfigMap"):
+		functionConfig.GetOrDie(&imageTrans.Image, "data")
+	case functionConfig.IsGVK(fn.KptFunctionGroup, fn.KptFunctionVersion, FnConfigKind):
+		functionConfig.GetOrDie(&imageTrans.Image, "image")
+		functionConfig.GetOrDie(&imageTrans.AdditionalImageFields, "additionalImageFields")
+	default:
+		return fmt.Errorf("unknown functionConfig Kind=%v ApiVersion=%v, expect `%v` or `ConfigMap` with correct formatting",
+			functionConfig.GetKind(), functionConfig.GetAPIVersion(), FnConfigKind)
+	}
+	if err := imageTrans.validateInput(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// validateInput validates the inputs passed into via the functionConfig
+func (imageTrans *ImageTransformer) validateInput() error {
+	if imageTrans.Image.Name == "" && imageTrans.Image.ContainerName == "" && imageTrans.Image.ImageName == "" {
+		return fmt.Errorf("missing image name or container name")
```

Comment:
How about `either `imageName` or `containerName` is required`. This helps users to navigate the exact function configs.